### PR TITLE
Make v3 discovery and etcdctl reuse the same ClientConfig

### PR DESCRIPTION
**Pull request by ahrtr**
Part of [13624](https://github.com/etcd-io/etcd/issues/13624), and it's one of the follow-ups to PR [13635](https://github.com/etcd-io/etcd/pull/13635).

Move the original [clientConfig](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L100) from `etcdctl` to `clientv3`, so that it can be reused by both etcdctl and v3 discovery. 

The original `clientConfig` has some dependencies on [pks/cobrautil](https://github.com/etcd-io/etcd/tree/main/pkg/cobrautl),  such as (https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L180). However, clientv3 isn't allowed to depend on `pkg`, see [go.mod#L50](https://github.com/etcd-io/etcd/blob/main/client/v3/go.mod#L50). Actually the original [pks/cobrautil](https://github.com/etcd-io/etcd/tree/main/pkg/cobrautl) is only used by `etcdctl` or `etcdutl`, so move it into the `clientv3` package as well.

Note that etcd uses golang's standard package `flag` to process command line flags, while etcdctl uses `github.com/spf13/pflag`, and etcdctl and v3 discovery has different flag names, such as `--endpoints` vs `--discovery-endpoints`.  So it's difficult to share the same code to process flags between them. So the command line flags keep unchanged in this PR. 

cc @ptabor @serathius @spzala

### CODE REVIEW

File: server/etcdserver/api/v3discovery/discovery.go [hunk 1]
```diff
@@ -55,22 +55,8 @@ var (
...
-
-	InsecureTransport  bool   `json:"discovery-insecure-transport"`
-	InsecureSkipVerify bool   `json:"discovery-insecure-skip-tls-verify"`
-	CertFile           string `json:"discovery-cert"`
-	KeyFile            string `json:"discovery-key"`
-	TrustedCAFile      string `json:"discovery-cacert"`
-
-	User     string `json:"discovery-user"`
-	Password string `json:"discovery-password"`
+	clientv3.ClientConfig `json:"client-config"`
```
Comment (serathius) [inline server/etcdserver/api/v3discovery/discovery.go:58]:
```suggestion
	clientv3.ClientConfig `json:"client"`
```
In hierarchy `Config -> DiscoveryConfig -> ClientConfig` there is no need to mention `config` in json keys as they are implicit.

File: client/v3/config.go [hunk 2]
```diff
@@ -90,3 +93,103 @@ type Config struct {
...
 	// TODO: support custom balancer picker
 }
+
+type ClientConfig struct {
+	Endpoints        []string      `json:"endpoints"`
+	RequestTimeOut   time.Duration `json:"request-timeout"`
+	DialTimeout      time.Duration `json:"dial-timeout"`
+	KeepAliveTime    time.Duration `json:"keepalive-time"`
+	KeepAliveTimeout time.Duration `json:"keepalive-timeout"`
+	Scfg             *SecureCfg    `json:"secure-cfg"`
```
Comment (serathius) [inline client/v3/config.go:103]:
```suggestion
	Secure             *SecureConfig    `json:"secure"`
```

File: client/v3/config.go [hunk 3]
```diff
@@ -90,3 +93,103 @@ type Config struct {
...
+	Acfg             *AuthCfg      `json:"auth-cfg"`
+}
+
+type SecureCfg struct {
```
Comment (serathius) [inline client/v3/config.go:107]:
```suggestion
type SecureConfig struct {
```
Let's be consistent in naming. If we have `ClientConfig` then let's use `SecureConfig` etc.

File: client/v3/config.go [hunk 4]
```diff
@@ -90,3 +93,103 @@ type Config struct {
...
+	Cert       string `json:"cert"`
+	Key        string `json:"key"`
+	Cacert     string `json:"cacert"`
+	ServerName string `json:"servername"`
+
+	InsecureTransport  bool `json:"insecure-transport"`
+	InsecureSkipVerify bool `json:"insecure-skip-tls-verify"`
+}
+
+type AuthCfg struct {
```
Comment (serathius) [inline client/v3/config.go:117]:
```suggestion
type AuthConfig struct {
```

File: client/v3/config.go [hunk 5]
```diff
@@ -90,3 +93,103 @@ type Config struct {
...
 }
+
+type ClientConfig struct {
+	Endpoints        []string      `json:"endpoints"`
+	RequestTimeOut   time.Duration `json:"request-timeout"`
+	DialTimeout      time.Duration `json:"dial-timeout"`
+	KeepAliveTime    time.Duration `json:"keepalive-time"`
+	KeepAliveTimeout time.Duration `json:"keepalive-timeout"`
+	Scfg             *SecureCfg    `json:"secure-cfg"`
+	Acfg             *AuthCfg      `json:"auth-cfg"`
```
Comment (serathius) [inline client/v3/config.go:104]:
```suggestion
	Auth             *AuthConfig      `json:"auth"`
```

File: client/v3/config.go [hunk 6]
```diff
@@ -90,3 +93,103 @@ type Config struct {
 
 	// TODO: support custom balancer picker
 }
+
+type ClientConfig struct {
```
Comment (serathius) [inline client/v3/config.go:97]: Note, let's avoid pointers in configuration. There is no benefit of using pointers here.

File: client/v3/config.go [hunk 7]
```diff
@@ -90,3 +93,103 @@ type Config struct {
...
+
+	client, err := New(*cfg)
+	if err != nil {
+		cobrautl.ExitWithError(cobrautl.ExitBadConnection, err)
+	}
+
+	return client
+}
+
+func NewClientCfg(endpoints []string, dialTimeout, keepAliveTime, keepAliveTimeout time.Duration, scfg *SecureCfg, acfg *AuthCfg) (*Config, error) {
```
Comment (serathius) [inline client/v3/config.go:136]: Would recommend constructor only be responsible for providing default values and not setting or validating. For configuration structs I expect values to be provided via `json.Unmarshall` or `flags.VarString`. Both of those functions take reference to struct/field making the constructor you wrote not very useful.

Example on how to define configuration (based on how K8s code does configuration):
```
type ClientConfig struct {
  DialTimeout time.Duration `json:"dial-timeout"`
}

func NewClientConfig() ClientConfig {
  return ClientConfig{
    DialTimeout: 5*time.Second,
  }
}

func (c ClientConfig) Validate() err {
  if c.DialTimeout < 0 {
    return fmt.Error("Timeout cannot be negative")
  }
  return nil
}

func (c ClientConfig) AddFlags(fs flags.FlagSet) {
  fs.VarString("dial-timeout", &c.DialTimeout, c.DialTimeout, "Dial timeout")
}

func main() {
  
  config := NewClientConfig()
  config.AddFlags(flags.CommandLine) // or something like
  err := config.Validate()
  if err != nil {
    fmt.Print(err)
    os.Exit(1)
  }
}
```

This code is very draftly, however shows overall direction. As for registering single `ClientConfig` as different flags we could use aliasing. For example:

```
func addClientConfigFlags(c ClientConfig, prefix string, fs flags.FlagSet) {
  fs.VarString(prefix + "-dial-timeout", &c.DialTimeout, c.DialTimeout, "Dial timeout")
}

type DiscoveryClientConfig ClientConfig
type EtcdctlClientConfig ClientConfig

func (c DialClientConfig) AddFlags(fs flags.FlagSet) {
  addClientConfigFlags(c, "discovery", fs)
}

func (c EtcdctlClientConfig) AddFlags(fs flags.FlagSet) {
  addClientConfigFlags(c, "client", fs)
}
```

**Comment by serathius** [comment]
This PR is little hard to read due to lot of dependency changes. Could you please create a separate PR/commiut that just fixes the dependencies? It should be fast and easy to merge, allowing us to focus the discussion on the config changes.

**Comment by ahrtr** [comment]
> This PR is little hard to read due to lot of dependency changes. Could you please create a separate PR/commiut that just fixes the dependencies? It should be fast and easy to merge, allowing us to focus the discussion on the config changes.

All the changes in go.mod and go.sum are made by` go mod tidy` automatically.  

Just as I mentioned in the description of this PR, I moved the package `pkg/corbautil` into clientv3 due to some reasons, and it causes some dependency changes. So I have to fix the dependency changes in this PR.

File: server/etcdserver/api/v3discovery/discovery.go (hunk 1 above)
Comment (ahrtr) [reply server/etcdserver/api/v3discovery/discovery.go:58]: makes sense. Resolved & thanks.

File: client/v3/config.go (hunk 6 above)
Comment (ahrtr) [reply client/v3/config.go:97]: I just moved the struct from etcdctl into clientv3.  I did not & do not want to change anything else so as not to impact the existing logic. If we want to change anything else, we'd better to do it in a separate PR. WDYT?

File: client/v3/config.go (hunk 2 above)
Comment (ahrtr) [reply client/v3/config.go:103]: fixed & thanks.

File: client/v3/config.go (hunk 3 above)
Comment (ahrtr) [reply client/v3/config.go:107]: fixed & thanks.

File: client/v3/config.go (hunk 4 above)
Comment (ahrtr) [reply client/v3/config.go:117]: fixed & thanks

File: client/v3/config.go (hunk 5 above)
Comment (ahrtr) [reply client/v3/config.go:104]: fixed & thanks.

File: client/v3/config.go (hunk 6 above)
Comment (serathius) [reply client/v3/config.go:97]: ack

File: client/v3/config.go [hunk 8]
```diff
@@ -90,3 +93,103 @@ type Config struct {
...
+	KeepAliveTimeout time.Duration `json:"keepalive-timeout"`
+	Secure           *SecureConfig `json:"secure"`
+	Auth             *AuthConfig   `json:"auth"`
+}
+
+type SecureConfig struct {
+	Cert       string `json:"cert"`
+	Key        string `json:"key"`
+	Cacert     string `json:"cacert"`
+	ServerName string `json:"servername"`
```
Comment (serathius) [inline client/v3/config.go:111]:
```suggestion
	ServerName string `json:"server-name"`
```

Comment (ahrtr) [reply client/v3/config.go:111]: Ah-:), I changed the name between "servername" and "server-name" several times back and forth. In the beginning, I used `server-name`, but I see "username", so I changed it back to "servername". 

Now you proposed "server-name" again, OK, then let me change it to "server-name".

**Comment by serathius** [comment]
> Just as I mentioned in the description of this PR, I moved the package `pkg/corbautil` into clientv3 due to some reasons, and it causes some dependency changes. So I have to fix the dependency changes in this PR.

Can you create a separate PR to move `pkg/corbautil` to `clientv3` then? It will make the review much much easier.

**Comment by ahrtr** [comment]
> > Just as I mentioned in the description of this PR, I moved the package `pkg/corbautil` into clientv3 due to some reasons, and it causes some dependency changes. So I have to fix the dependency changes in this PR.
> 
> Can you create a separate PR to move `pkg/corbautil` to `clientv3` then? It will make the review much much easier.

Actually I want to do it in a separate PR as well, but I can't. I mentioned the reason in the description of this PR. Pasted it again. The original `clientConfig` has some dependencies on [pks/cobrautil](https://github.com/etcd-io/etcd/tree/main/pkg/cobrautl), such as (https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L180). However, clientv3 isn't allowed to depend on pkg, see [go.mod#L50](https://github.com/etcd-io/etcd/blob/main/client/v3/go.mod#L50). Actually the original [pks/cobrautil](https://github.com/etcd-io/etcd/tree/main/pkg/cobrautl) is only used by etcdctl or etcdutl, so move it into the clientv3 package as well.

**Comment by serathius** [comment]
I don't think that tools for cobra should be put into client package. We should not put dependency on cobra for users that just want to call etcd API. I think we should update the code to stop using cobrautil as dependency.

**Comment by ahrtr** [comment]
Actually it just depends on [cobrautl.ExitBadArgs](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L180) and [cobrautl.ExitBadConnection](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L180). 

If you are OK, I can remove the dependency in this PR by change them with the related number directly in this PR, see below,

```
cobrautl.ExitWithError(128, err)

cobrautl.ExitWithError(2, err)
```

Then I can update all the dependency related changes in a separate PR.  WDYT? @serathius

**Comment by serathius** [comment]
I think that while trying to reuse ClientConfig we are stumbling upon `library` vs `binary` capabilities. Here the difference is seen in error handling. As binary know what command line framework it uses `cobra` and knows what error codes it want to return it can use them directly. This is however untrue for libraries that should return error to user and let them decide if program should exit or not. 

By reusing `ClientConfig` we are trying it to become a library, however the code itself needs to be changed. I think that instead of calling the exit codes directly we should return error that would be used by binary to decide which error code to return.

**Pull request closed without merging**